### PR TITLE
AutoLock, Application Preferences and more

### LINKED ISSUE #41: Add user preference for window.setFlags(... FLAG_SECURE)
Create a preferences feature that allows the user to disable screen protection (FLAG_SECURE).

```
window?.setFlags(
    WindowManager.LayoutParams.FLAG_SECURE,
    if (userHasDisabledScreenProtection) 0 else WindowManager.LayoutParams.FLAG_SECURE
)
```

Do not allow the user to set the preference to turn FLAG_SECURE off without reading this message

> Do not be tricked taking screenshots or recording images of your DiceKey. Anyone who sees the image of your key can replicate it.  Enable this feature only when using sample DiceKeys, such as if recording a demo video or product review. To allow recording, type "I know any key I record will be compromised" in the field below.

### PR DESCRIPTION
**Pull request by angelix**
Prevent screenshots and screen capturing (configurable)
AutoLock DiceKeys after x minutes
Fix #41, #79

### CODE REVIEW

File: app/src/main/java/org/dicekeys/app/fragments/dicekey/AbstractDiceKeyFragment.kt [hunk 1]
```diff
@@ -21,26 +22,62 @@ import javax.inject.Inject
...
+
+    // Check if DiceKey is still in memory
+    private fun guard(){
         // Guard: If DiceKey is not available, return
         repository.getActiveDiceKey()?.also {
             diceKey = it
         } ?: run {
             findNavController().popBackStack()
-            toast("DiceKey not found in memory")
+            // toast("DiceKey not found in memory")
```
Comment (UppaJung) [inline app/src/main/java/org/dicekeys/app/fragments/dicekey/AbstractDiceKeyFragment.kt:79]: Delete entirely?

File: app/src/main/java/org/dicekeys/app/viewmodels/ListDiceKeysViewModel.kt [hunk 2]
```diff
@@ -10,8 +10,15 @@ import javax.inject.Inject
 
 @HiltViewModel
 class ListDiceKeysViewModel @Inject constructor(private val encryptedStorage: EncryptedStorage, private val diceKeyRepository: DiceKeyRepository) : ViewModel() {
+
+    fun isDiceKeyInMemory(encryptedDiceKey: EncryptedDiceKey) = diceKeyRepository.exists(encryptedDiceKey)
+
+    fun getDiceKey(encryptedDiceKey: EncryptedDiceKey)  = diceKeyRepository.get(encryptedDiceKey.keyId)
+
     fun remove(encryptedDiceKey: EncryptedDiceKey){
+        // Remove from Storage
```
Comment (UppaJung) [inline app/src/main/java/org/dicekeys/app/viewmodels/ListDiceKeysViewModel.kt:19]: In what context would the user ask to remove a DiceKey from both encrypted long-term storage and memory at once?

**Review by UppaJung — APPROVED** [review]
See comments.

**Comment by UppaJung** [comment]
This appears to takes a different approach to auto-locking than the iOS/macOS approach sent for review 6 days ago:

https://github.com/dicekeys/dicekeys-app-ios/pull/122/files

I didn't document the the iOS/macOS approach well for replication.  It starts the timer when one leaves the view of a specific DiceKey and goes to the main view.  It then displays a countdown.

The problem with the iOS/macOS approach is that the key will stay in memory indefinitely if the user navigates away in one of the views where the countdown is halted (any view other than the main view).

I don't fully understand the policy for Android here.  We should document both and try to resolve to make them consistent.

**Comment by angelix** [comment]
> This appears to takes a different approach to auto-locking than the iOS/macOS approach sent for review 6 days ago:
> 
> https://github.com/dicekeys/dicekeys-app-ios/pull/122/files
> 
> I didn't document the the iOS/macOS approach well for replication. It starts the timer when one leaves the view of a specific DiceKey and goes to the main view. It then displays a countdown.
> 
> The problem with the iOS/macOS approach is that the key will stay in memory indefinitely if the user navigates away in one of the views where the countdown is halted (any view other than the main view).
> 
> I don't fully understand the policy for Android here. We should document both and try to resolve to make them consistent.

I tried to keep it simple it as much as possible. My solution was based on [ProcessLifecycleOwner](https://developer.android.com/reference/androidx/lifecycle/ProcessLifecycleOwner), an easy/recommended way to detect when the app is on background/foreground.

If the app goes to the background, after X seconds/minutes (based on user preferences, default value 1 minute), *all* DiceKeys are removed from memory.
When the app goes to foreground, the timer stops and resets.

No visual timer, as the timer is initiated only after the app goes to the background.

**Pull request merged**